backend/packages/Upgrade/src/api/middlewares/UserCheckMiddleware.ts

```diff
@@ -38,4 +46,159 @@ export class UserCheckMiddleware {
+   *   context: "storedUserDependentContext",
+   * }
+   *
+   * // Scenario 2: Ignore stored user and use session groups only
+   * {
+   *   context: "independentContext",
+   *   groupsForSession: { "classId": ["testClass"], "schoolId": ["instructor"] },
+   *   includeStoredUserGroups: false
+   * }
+   *
+   * // Scenario 3: Merge session and stored groups
+   * {
+   *   context: "storedUserMergedContext",
+   *   groupsForSession: { "classId": ["testClass"], "schoolId": ["demoSchool"] },
+   *   includeStoredUserGroups: true
+   * }
+   *
+   * ```
+   */
+  private async handleProvidedGroupsForSession(req: AppRequest, user_id: string): Promise<RequestedExperimentUser> {
+    // Note: validation of request will have occurred before this middleware
+
+    // Scenario 1: Session-only groups (ephemeral user mode)
+    // explicitly check if includeStoredUserGroups is exactly false and not just undefined
+    if (req.body.groupsForSession && req.body.includeStoredUserGroups === false) {
+      req.logger.info({
+        message: 'Using session-only groups (ephemeral user mode)',
+      });
+
+      return this.createSessionUser(user_id, req.body.groupsForSession);
+    }
+
+    // Load stored user document for scenarios 2 and 3
+    const experimentUserDoc = await this.experimentUserService.getUserDoc(user_id, req.logger);
+
+    // Scenario 2: Merged groups (Merged stored/ephemeral groups mode)
+    if (req.body.includeStoredUserGroups && req.body.groupsForSession) {
+      if (experimentUserDoc?.group) {
+        experimentUserDoc.group = this.mergeGroupsWithUniqueValues(experimentUserDoc.group, req.body.groupsForSession);
+
+        req.logger.info({
+          message: 'Merged session groups with stored user groups',
+          experimentUserDoc,
+        });
+      } else {
+        req.logger.info({
+          message: 'No stored groups found, using session groups only',
+        });
+
+        // If no stored groups exist, just use session groups
+        return this.createSessionUser(user_id, req.body.groupsForSession);
+      }
+    } else {
+      // Scenario 3: Standard behavior (user-lookup from user-id)
+      req.logger.debug({
+        message: 'Using standard user lookup without session group modifications',
+        experimentUserDoc,
+      });
+    }
+
+    return experimentUserDoc;
+  }
+
+  /**
+   * Creates a RequestedExperimentUser object for session-only or ephemeral user scenarios.
+   *
+   * @param user_id - The user ID for the experiment user
+   * @param groupsForSession - The groups provided in the session
+   * @returns RequestedExperimentUser object with session groups
+   */
+  private createSessionUser(user_id: string, groupsForSession: Record<string, string[]>): RequestedExperimentUser {
+    const sessionUser = new RequestedExperimentUser();
+    sessionUser.id = user_id;
+    sessionUser.requestedUserId = user_id;
+    sessionUser.group = groupsForSession;
+    sessionUser.workingGroup = undefined;
```

Comment:
Users with no working group information will now always be excluded from feature flags in contexts where there's a global exclude defined with at least one group-type list. I think I recall that we were ok with that for now, and there's still stuff to be worked out w.r.t feature flags and global excludes?

Reply:
yeah should be ok, esp. considering we know how this will be used initially, but good to remember for sure.